### PR TITLE
Avoid producing tons of state maps for single node cursor

### PR DESCRIPTION
**Pull request by MishaDemianenko**
Remember state of already defined nodes only for case of nodes iteration
and do not do that if cursor created to go over single node.

### CODE REVIEW

**Comment by MishaDemianenko** [comment]
@tinwelint @pontusmelke guys, can you please take a look, it resolves super annoying slowness problem

File: community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/NodeCursor.java [hunk 1]
```diff
@@ -91,7 +94,8 @@ void single( long reference, Read read )
         this.highMark = NO_ID;
         this.read = read;
         this.hasChanges = HasChanges.MAYBE;
-        this.addedNodes = null;
+        this.addedNodes = emptySet();
+        this.single = true;
```
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/NodeCursor.java]: I think you can use `this.highMark == NO_ID` to check for `single` without introducing a new field. There is already a method `isSingle` method you can use instead.

**Review by tinwelint — APPROVED** [review]
Makes it slightly more complicated, but nice improvement imho

File: community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/NodeCursor.java (hunk 1 above)
Comment (MishaDemianenko) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/NodeCursor.java]: Thx! was blind or something. Switch to use that method instead.

**Pull request merged**